compiler/p/codegen/OMRCodeGenerator.cpp

```diff
@@ -2415,8 +2415,10 @@ int32_t OMR::Power::CodeGenerator::getMaximumNumberOfGPRsAllowedAcrossEdge(TR::N
          if (total <= 9)
             {
             for (ii=3; ii<total &&
-                       (node->getChild(ii)->getCaseConstant()-node->getChild(ii-1)->getCaseConstant())<=UPPER_IMMED; ii++)
-            ;
+                       (node->getChild(ii)->getCaseConstant()-node->getChild(ii-1)->getCaseConstant())<=UPPER_IMMED &&
+                       (node->getChild(ii)->getCaseConstant()-node->getChild(ii-1)->getCaseConstant())>0; /* in case of an overflow */
```

Comment:
There are 3 types of lookup schemes used in switchDispatch():

1. All cases constants can fit into a 16bit int
2. The stride between any two adjacent cases can fit into a 16bit int
3. A scheme for scenarios that don't fit into #1 or #2 above.

The number of registers that we allow GRA to use (allowGlobalRegisterAcrossBranch()) across the switch code is different for each case because the number of registers needed to generate the switch control flow is different. 

The problem here has to do with the #2 scenario. When calculating the difference, if a integer overflow occurs the "difference<=UPPER_IMMED" test will pass despite the fact that the difference can't truly fit in a 16bit int.

The logic in switchDispatch() is checking for the overflow when deciding which lookup scheme to use, but the logic in allowGlobalRegisterAcrossBranch() which controls how many registers are made available to GRA does not check for an overflow, this fix adds the missing overflow check so that the allowGlobalRegisterAcrossBranch() logic matches the logic in switchDispatch().

Reply:
Thanks!